[distributed] Self-adaptive `send()`/`recv()` pair

sjiang95 opened the pull request:
Currently, the Point-to-point communication pair
`torch.distributed.send()` and `torch.distributed.recv()`
requires the user to provide `recv()` a tensor of the same
`shape` and `dtype` with that in corresponding `send()`.

In practice, the shape and dtype of src tensor in `send()` can be hard
or just inconvenient to know, or they changes during time.

This PR implements `adaptive_send()`/`adaptive_recv()` pair.
`adaptive_recv()` requires no input tensor as container, but returns a
new tensor whose `shape` and `dtype` is decided by the src tensor sent
by `adaptive_send()`.

Related issue: #81545

Signed-off-by: Shengjiang QUAN <qsj287068067@126.com>

Fixes #81545


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

AdorableJiang commented:
## A minimum demo

```python
import torch
import torch.multiprocessing as mp
import torch.distributed as dist
import time
import random

def main_worker(rank, world_size):
    dist.init_process_group(
        backend="nccl",
        init_method="tcp://127.0.0.1:9001",
        world_size=world_size,
        rank=rank,
    )
    print("process begin", rank)

    # adaptive_ send()/recv()
    # test float tensors on adaptive_send()/adaptive_recv()
    for datatype in [torch.float16, torch.float, torch.float32, torch.float64, torch.bfloat16]:#torch.int8, torch.int32, torch.int64, 
        if rank == 0:
            print(f"Current datatype: {datatype}.")
            t = torch.rand([random.randint(1,5),random.randint(1,5),random.randint(1,2)],dtype=datatype).squeeze()
            t=t.to(torch.device('cuda',rank))
            print(f"send {t.size()} tensor\n{t}")
            dist.adaptive_send(t,dst=1,tag=1)   

        elif rank == 1:
            _, r = dist.adaptive_recv(src=0, tag=1)#
            print(f"recv {r.size()} tensor\n{r}")
        print()
        time.sleep(1)

    # test int tensors on adaptive_send()/adaptive_recv()
    for datatype in [torch.int8, torch.int32, torch.int64]:#
        if rank == 0:
            print(f"Current datatype: {datatype}.")
            t = torch.randint(high=100,size=[random.randint(1,5),random.randint(1,5),random.randint(1,2)],dtype=datatype).squeeze()
            t=t.to(torch.device('cuda',rank))
            print(f"send {t.size()} tensor\n{t}")
            dist.adaptive_send(t,1,tag=1) 

        elif rank == 1:
            _, r=dist.adaptive_recv(0,tag=1)
            print(f"recv {r.size()} tensor\n{r}")
        print()
        time.sleep(1)

def main():
    mp.spawn(main_worker, nprocs=2, args=(2,))

if __name__ == "__main__":
    main()
```

## Output

<table>
<tr>
	<td> rank0
	<td> rank1
<tr>
<td>
	
```python
process begin 0
Current datatype: torch.float16.
send torch.Size([]) tensor
0.91943359375

Current datatype: torch.float32.
send torch.Size([5, 4]) tensor
tensor([[0.6875, 0.2779, 0.0606, 0.6943],
        [0.9978, 0.8599, 0.7083, 0.5698],
        [0.4394, 0.8576, 0.7520, 0.1517],
        [0.8000, 0.4255, 0.7674, 0.3102],
        [0.7461, 0.1194, 0.4881, 0.4013]], device='cuda:0')

Current datatype: torch.float32.
send torch.Size([2]) tensor
tensor([0.8165, 0.9932], device='cuda:0')

Current datatype: torch.float64.
send torch.Size([3, 4, 2]) tensor
tensor([[[0.7860, 0.8269],
         [0.8732, 0.5874],
         [0.5057, 0.8407],
         [0.4018, 0.7303]],

        [[0.4590, 0.1504],
         [0.7995, 0.4520],
         [0.8155, 0.0474],
         [0.5773, 0.0664]],

        [[0.5335, 0.2403],
         [0.8872, 0.2634],
         [0.4001, 0.2514],
         [0.2262, 0.4040]]], device='cuda:0', dtype=torch.float64)

Current datatype: torch.bfloat16.
send torch.Size([5, 4, 2]) tensor
tensor([[[0.3633, 0.7500],
         [0.5391, 0.8008],
         [0.0391, 0.0117],
         [0.6953, 0.4062]],

        [[0.2891, 0.1992],
         [0.7344, 0.9883],
         [0.7422, 0.2344],
         [0.2734, 0.2188]],

        [[0.6992, 0.3672],
         [0.3984, 0.2852],
         [0.5039, 0.7500],
         [0.1797, 0.6641]],

        [[0.9688, 0.9531],
         [0.1836, 0.7266],
         [0.0586, 0.2578],
         [0.2422, 0.0508]],

        [[0.7852, 0.0469],
         [0.0742, 0.9531],
         [0.8086, 0.0742],
         [0.8320, 0.1211]]], device='cuda:0', dtype=torch.bfloat16)

Current datatype: torch.int8.
send torch.Size([3]) tensor
tensor([34, 55, 67], device='cuda:0', dtype=torch.int8)

Current datatype: torch.int32.
send torch.Size([4, 4]) tensor
tensor([[91, 80, 42, 30],
        [78, 18, 54, 89],
        [39, 85,  1, 13],
        [59, 90, 36, 65]], device='cuda:0', dtype=torch.int32)

Current datatype: torch.int64.
send torch.Size([4, 2]) tensor
tensor([[12, 78],
        [63, 13],
        [42, 86],
        [81, 53]], device='cuda:0')
```

</td>
<td>

```python
process begin 1
recv torch.Size([]) tensor
0.91943359375

recv torch.Size([5, 4]) tensor
tensor([[0.6875, 0.2779, 0.0606, 0.6943],
        [0.9978, 0.8599, 0.7083, 0.5698],
        [0.4394, 0.8576, 0.7520, 0.1517],
        [0.8000, 0.4255, 0.7674, 0.3102],
        [0.7461, 0.1194, 0.4881, 0.4013]], device='cuda:1')

recv torch.Size([2]) tensor
tensor([0.8165, 0.9932], device='cuda:1')

recv torch.Size([3, 4, 2]) tensor
tensor([[[0.7860, 0.8269],
         [0.8732, 0.5874],
         [0.5057, 0.8407],
         [0.4018, 0.7303]],

        [[0.4590, 0.1504],
         [0.7995, 0.4520],
         [0.8155, 0.0474],
         [0.5773, 0.0664]],

        [[0.5335, 0.2403],
         [0.8872, 0.2634],
         [0.4001, 0.2514],
         [0.2262, 0.4040]]], device='cuda:1', dtype=torch.float64)

recv torch.Size([5, 4, 2]) tensor
tensor([[[0.3633, 0.7500],
         [0.5391, 0.8008],
         [0.0391, 0.0117],
         [0.6953, 0.4062]],

        [[0.2891, 0.1992],
         [0.7344, 0.9883],
         [0.7422, 0.2344],
         [0.2734, 0.2188]],

        [[0.6992, 0.3672],
         [0.3984, 0.2852],
         [0.5039, 0.7500],
         [0.1797, 0.6641]],

        [[0.9688, 0.9531],
         [0.1836, 0.7266],
         [0.0586, 0.2578],
         [0.2422, 0.0508]],

        [[0.7852, 0.0469],
         [0.0742, 0.9531],
         [0.8086, 0.0742],
         [0.8320, 0.1211]]], device='cuda:1', dtype=torch.bfloat16)

recv torch.Size([3]) tensor
tensor([34, 55, 67], device='cuda:1', dtype=torch.int8)

recv torch.Size([4, 4]) tensor
tensor([[91, 80, 42, 30],
        [78, 18, 54, 89],
        [39, 85,  1, 13],
        [59, 90, 36, 65]], device='cuda:1', dtype=torch.int32)

recv torch.Size([4, 2]) tensor
tensor([[12, 78],
        [63, 13],
        [42, 86],
        [81, 53]], device='cuda:1')
```

</td>
</table>

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
+    "torch.float32" : 7, # ncclFloat32    = 7
+    "torch.float64" : 8, # ncclFloat64    = 8
+    "torch.bfloat16": 9, # ncclBfloat16   = 9
+}
```
Is there a reason to introduce NCCL specific coding here? Same below.

AdorableJiang replied:
No special reason. Just need a way to encode the `dtype`. For overall consistency of `PyTorch` source code, I choosed to use the NCCL specific coding here.

AdorableJiang replied:
Just need a mapping to encode tensor dtypes. I have changed that to `ScalarType`.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
+
+    if group is None or group is GroupMember.WORLD:
+        default_pg = _get_default_group()
+        default_pg.send([send_len_tensor_info], dst, tag=tag-1).wait()
```
Is it safe to use tag-1? What if a previous operation is using tag-1?

AdorableJiang replied:
This is also my main concern. Do you have any suggestion on this?

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
+        default_pg = _get_default_group()
+        default_pg.send([send_len_tensor_info], dst, tag=tag-1).wait()
+        # send `tensor`'s `dtype` first
+        default_pg.send([send_tensor_info], dst, tag=tag+1).wait()
```
Same here for tag+1

AdorableJiang replied:
Same above.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
+        pg = group
+
+    recv_len_tensor_info=torch.tensor([-1],dtype=torch.int).cuda(get_rank())
+    recv_tensor_dtype=torch.tensor([-1],dtype=torch.int).cuda(get_rank())
```
Not sure why the tensor's residence has to be CUDA. What if there is no CUDA support? Same below.

kwen2501 replied:
Also, `get_rank()` would return the rank of the current process in the entire group rather than a local CUDA device ID.

sjiang95 committed:
[distributed] Self-adaptive `send()`/`recv()` pair

Currently, the Point-to-point communication pair
`torch.distributed.send()` and `torch.distributed.recv()`
requires the user to provide `recv()` a tensor of the same
`shape` and `dtype` with that in corresponding `send()`.

In practice, the shape and dtype of src tensor in `send()` can be hard
or just inconvenient to know, or they changes during time.

This PR implements `adaptive_send()`/`adaptive_recv()` pair.
`adaptive_recv()` requires no input tensor as container, but returns a
new tensor whose `shape` and `dtype` is decided by the src tensor sent
by `adaptive_send()`.

Related issue: #81545

Signed-off-by: Shengjiang QUAN <qsj287068067@126.com>

AdorableJiang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1064,6 +1064,165 @@ def recv(tensor, src=None, group=None, tag=0):
+    else:
+        pg = group
+
+    current_device=_get_pg_device(group=pg) # cuda for NCCL and CPU for everything else
```
Detect device type.

AdorableJiang commented:
@kwen2501 May I invite you to see this again?
